### PR TITLE
Using ETEXT_DESC for status update

### PR DESCRIPTION
**Pull request by chrizztus**
Hi,
for my project I needed the recognizing status. Within the headers I found  ETEXT_DESC which exactly does what I need. You just need to pass it's pointer to `_tesseract->Recognize()`.
In order to update the value I created a thread which periodically checks the progress field of ETEXT_DESC for changes.
Maybe you'd like to merge into your master branch after testing. For me it works quiet well.

### CODE REVIEW

File: Classes/Tesseract.mm [hunk 1]
```diff
@@ -35,6 +36,8 @@ @interface Tesseract () {
 
 @implementation Tesseract
 
+@synthesize tessDelegate;
```
Comment (ldiqual) [inline Classes/Tesseract.mm:39]: This is unnecessary.

File: Classes/Tesseract.mm [hunk 2]
```diff
@@ -192,6 +195,40 @@ - (BOOL)recognize {
...
 }
+  
+- (BOOL)recognizeWithProgressUpdate {
+
+    _progress = 0;
+    ETEXT_DESC *monitor = new ETEXT_DESC;
+
+    void (^progressBlock)(void) = ^void
+    {
+        if (monitor->progress > _progress)
```
Comment (ldiqual) [inline Classes/Tesseract.mm]: `) {`

File: Classes/Tesseract.mm [hunk 3]
```diff
@@ -192,6 +195,40 @@ - (BOOL)recognize {
     int returnCode = _tesseract->Recognize(NULL);
     return (returnCode == 0) ? YES : NO;
 }
+  
+- (BOOL)recognizeWithProgressUpdate {
+
+    _progress = 0;
+    ETEXT_DESC *monitor = new ETEXT_DESC;
+
+    void (^progressBlock)(void) = ^void
```
Comment (ldiqual) [inline Classes/Tesseract.mm]: `^void {`

File: Classes/Tesseract.mm [hunk 4]
```diff
@@ -192,6 +195,40 @@ - (BOOL)recognize {
...
+    {
+        if (monitor->progress > _progress)
+        {
+            _progress = monitor->progress;
+            [self.tessDelegate progressUpdate:_progress];
+        }
+    };
+  
+    dispatch_queue_t backgroundQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
+    dispatch_source_t timer = createDispatchTimer(1,
```
Comment (ldiqual) [inline Classes/Tesseract.mm]: One line should be enough.

File: Classes/Tesseract.mm [hunk 5]
```diff
@@ -192,6 +195,40 @@ - (BOOL)recognize {
...
+                                                  0,
+                                                  backgroundQueue,
+                                                  progressBlock);
+  
+    int returnCode = _tesseract->Recognize(monitor);
+
+    if(_progress < 100)
```
Comment (ldiqual) [inline Classes/Tesseract.mm]: `if (` and `) {`

File: Classes/Tesseract.mm [hunk 6]
```diff
@@ -238,4 +275,23 @@ - (void)setImage:(UIImage *)image
     _tesseract->SetImage((const unsigned char *) _pixels, width, height, sizeof(uint32_t), width * sizeof(uint32_t));
 }
 
+#pragma mark - progress update
+
+dispatch_source_t createDispatchTimer(uint64_t interval,
+                                      uint64_t leeway,
+                                      dispatch_queue_t queue,
+                                      dispatch_block_t block)
```
Comment (ldiqual) [inline Classes/Tesseract.mm:277]: `) {`

File: Classes/Tesseract.mm [hunk 7]
```diff
@@ -238,4 +275,23 @@ - (void)setImage:(UIImage *)image
...
+{
+    dispatch_source_t timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER,
+                                                   0, 0, queue);
+    if (timer)
```
Comment (ldiqual) [inline Classes/Tesseract.mm]: `) {`

File: README.md [hunk 8]
```diff
@@ -31,6 +31,7 @@ Usage
 Here is the default workflow to extract text from an image:
 
  - Instantiate Tesseract with data path and language
+ - Set the delegate
```
Comment (ldiqual) [inline README.md]: Can you specify that it's optional?

File: README.md [hunk 9]
```diff
@@ -44,6 +45,7 @@ Code Sample
     #import "Tesseract.h"
     
     Tesseract* tesseract = [[Tesseract alloc] initWithDataPath:@"tessdata" language:@"eng"];
+    [tesseract setTessDelegate:self];
```
Comment (ldiqual) [inline README.md]: Can you add a code sample just for recognition with progress update? Let's not crowd the main example too much.

**Comment by ldiqual** [comment]
That's an awesome improvement! Thanks so much for contributing.
I added a few comments, mainly about style. I'll merge this PR if you fix them.

File: Classes/Tesseract.mm (hunk 1 above)
Comment (chrizztus) [inline Classes/Tesseract.mm:39]: deleted: see my latest commit [c72328a]

File: Classes/Tesseract.mm (hunk 2 above)
Comment (chrizztus) [inline Classes/Tesseract.mm]: changed: see my latest commit [c72328a]

File: Classes/Tesseract.mm (hunk 3 above)
Comment (chrizztus) [inline Classes/Tesseract.mm]: changed: see my latest commit [c72328a]

File: Classes/Tesseract.mm (hunk 4 above)
Comment (chrizztus) [inline Classes/Tesseract.mm]: changed: see my latest commit [c72328a]

File: Classes/Tesseract.mm (hunk 5 above)
Comment (chrizztus) [inline Classes/Tesseract.mm]: changed: see my latest commit [c72328a]

File: Classes/Tesseract.mm (hunk 6 above)
Comment (chrizztus) [inline Classes/Tesseract.mm:277]: changed: see my latest commit [c72328a]

File: Classes/Tesseract.mm (hunk 7 above)
Comment (chrizztus) [inline Classes/Tesseract.mm]: changed: see my latest commit [c72328a]

File: README.md (hunk 8 above)
Comment (chrizztus) [inline README.md]: see commit [486f077]

File: README.md (hunk 9 above)
Comment (chrizztus) [inline README.md]: added a new section in README.md for using recognize with progress update
see commit [c72328a]

**Comment by chrizztus** [comment]
Hi,
I added a few comments regarding my latest changes.
Hope everything is fine now.
Cheers

**Pull request closed without merging**